grpclog/component.go

```diff
@@ -0,0 +1,289 @@
+	TRANSPORT  = Component("TRANSPORT")
+	XDS        = Component("XDS")
+)
+
+var components = map[Component]*componentData{}
+var grpcComponents = []Component{
+	CORE,
+	ALTS,
+	GRPCLB,
+	RLS,
+	ROUNDROBIN,
+	BINARYLOG,
+	CHANNELZ,
+	DNS,
+	TRANSPORT,
+	XDS,
+}
+var environmentVars = map[string]*componentData{}
+var prefixVars = map[string]*componentData{}
+
+// init creates all grpc components and any components named in the environment variable and applies all settings
+// specified in the environment variable. Notably, settings for all components with a given prefix are applies before
+// regular component settings.
+func init() {
+	// Initialize the grpc components
+	for _, c := range grpcComponents {
+		components[c] = &componentData{0, 0}
+	}
+	// Pull environment variable data
+	if envVar, ok := os.LookupEnv(envName); ok && len(envVar) > 0 {
+		varList := strings.Split(envVar, ",")
+		for _, varPair := range varList {
+			varPairList := strings.Split(varPair, ":")
+			if len(varPairList) != 2 {
+				fmt.Fprintf(os.Stderr, "error: could not parse %v value '%v', unrecognized key-value pair '%v'\n", envName, envVar, varPair)
+				os.Exit(1)
+			}
+			if cData, ok := parseVar(varPairList[1]); ok {
+				if varPrefix, ok := getPrefix(varPairList[0]); ok {
+					prefixVars[varPrefix] = &componentData{cData.verbosity, cData.level}
+				} else {
+					environmentVars[varPairList[0]] = &componentData{cData.verbosity, cData.level}
+				}
+			} else {
+				fmt.Fprintf(os.Stderr, "error: could not parse %v value '%v', unrecognized value '%v'\n", envName, envVar, varPairList[1])
+				os.Exit(1)
+			}
+		}
+	}
+	// Create any components named in the environment variable that do not exist
+	for name := range environmentVars {
+		c := Component(name)
+		if _, ok := components[c]; !ok {
+			components[c] = &componentData{0, 0}
+		}
+	}
+	// Apply prefixes
+	for prefix, pData := range prefixVars {
+		for c, cData := range components {
+			if strings.HasPrefix(string(c), prefix) {
+				cData.apply(pData)
+			}
+		}
+	}
+	// Apply non-prefixes
+	for name, vData := range environmentVars {
+		c := Component(name)
+		if cData, ok := components[c]; ok {
+			cData.apply(vData)
+		}
+	}
+}
+
+// apply applies the parameter componentData to the receiver componentData.
+// Sentinel values in the parameter will not be applied.
+func (cData *componentData) apply(applyData *componentData) {
+	if applyData.verbosity != sentinel {
+		cData.verbosity = applyData.verbosity
+	}
+	if applyData.level != sentinel {
+		cData.level = applyData.level
+	}
+}
+
+// parseVar parses the value in a key-value pair from the environment variable.
+// The resulting componentData will have a sentinel value for the verbosity if it is unspecified in the value.
+// Second return value is false if there was an error parsing.
+func parseVar(s string) (componentData, bool) {
+	result := componentData{sentinel, sentinel}
+	switch {
+	case s == "INFO":
+		result.level = 0
```

Comment:
Please define consts instead of "0"/"1" etc.

Reply:
Made all log levels constants.